### PR TITLE
Add utility method Client#getClientUri

### PR DESCRIPTION
**Pull request by Scrumplex**
Hello there,
I've added a utility method to the `Client` class. It will return an URI, that is used by TeamSpeak 3 to reference a client. You could write that URI into a channel's description or just message it via chat.

Example URI:
`[URL=client://38/SFWII7e3mzgbS5ZRKKRq/vxm2i0=~Scrumplex]Scrumplex[/URL]`

You can get your client URI by draging your client into the chat box in your client. 
That URI is connection specific. Because it contains the client id.

### CODE REVIEW

**Comment by Firedroide** [comment]
Heyo!

Interesting pull request! I've never even heard of client URIs :smile:

There are a few issues, though, that I'd like to see addressed before I can merge this:

1. You changed all tabs to spaces, which makes the diff very hard to read. I'd prefer if you used tabs everywhere.
2. I don't see why the URL should already come inside `[URL]` BB-codes. It might be better if we'd just return the actual URL part to stay as "general-purpose" as possible.
3. A few minor things that I'll put into line comments.

Thanks for the PR :smile:

File: .gitignore [hunk 1]
```diff
@@ -14,3 +14,6 @@
 .settings
 /.apt_generated
 /html
+
+# Idea Files
+/.idea
```
Comment (Firedroide) [inline .gitignore]: Missing newline at end of file.

(I know this .gitignore doesn't really belong to this PR, but I'm fine with it being here as IntellJ's files should also be ignored anyways. So leave it in, but please fix the newline ^^)

File: src/main/java/com/github/theholywaffle/teamspeak3/api/wrapper/Client.java [hunk 2]
```diff
@@ -28,174 +28,201 @@
...
+        sb.append(getId());
+        sb.append("/");
+        sb.append(getUniqueIdentifier());
+        sb.append("~");
+        String encodedNick;
+        try {
+            //We will encode the nickname, because there are people, that are using incompatible chars in names.
+            encodedNick = URLEncoder.encode(getNickname(), "UTF-8"); //"The World Wide Web Consortium Recommendation states that UTF-8 should be used. Not doing so may introduce incompatibilites." Source: https://docs.oracle.com/javase/7/docs/api/java/net/URLEncoder.html#encode(java.lang.String,%20java.lang.String)
+        } catch (UnsupportedEncodingException e) {
+            encodedNick = getNickname();
```
Comment (Firedroide) [inline src/main/java/com/github/theholywaffle/teamspeak3/api/wrapper/Client.java]: Just re-throw the exception as a `new RuntimeException(e)`. UTF-8 is one of the charsets that every implementation of Java must support. Returning something that's probably-most-of-the-time-correct is not an option in this case.

File: src/main/java/com/github/theholywaffle/teamspeak3/api/wrapper/Client.java [hunk 3]
```diff
@@ -28,174 +28,201 @@
...
+    public String getClientUri() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("[URL=client://");
+        sb.append(getId());
+        sb.append("/");
+        sb.append(getUniqueIdentifier());
+        sb.append("~");
+        String encodedNick;
+        try {
+            //We will encode the nickname, because there are people, that are using incompatible chars in names.
```
Comment (Firedroide) [inline src/main/java/com/github/theholywaffle/teamspeak3/api/wrapper/Client.java]: This is **extremely** nitpicky, feel free to ignore this comment in its entirety :P

- People are not objects, so it should be "people who", not "people, that"
- There should be a space after `//`
- This whole try statement could be moved to the top. That way we'd first initialize all variables we're going to need and then use them.

File: src/main/java/com/github/theholywaffle/teamspeak3/api/wrapper/Client.java [hunk 4]
```diff
@@ -28,174 +28,201 @@
...
+            encodedNick = URLEncoder.encode(getNickname(), "UTF-8"); //"The World Wide Web Consortium Recommendation states that UTF-8 should be used. Not doing so may introduce incompatibilites." Source: https://docs.oracle.com/javase/7/docs/api/java/net/URLEncoder.html#encode(java.lang.String,%20java.lang.String)
```
Comment (Firedroide) [inline src/main/java/com/github/theholywaffle/teamspeak3/api/wrapper/Client.java]: The comment on this statement doesn't really add anything and goes way off screen. Consider removing it.

File: src/main/java/com/github/theholywaffle/teamspeak3/api/wrapper/Client.java [hunk 5]
```diff
@@ -28,174 +28,201 @@
...
+    /**
+     * Utility method. This method will return a client uri.
+     * A client uri can be used to reference a client.
+     * Example: [URL=client://5/thisisanuid=~Jesus]Jesus[/URL]
+     *
+     * @return clientUri
+     */
+    public String getClientUri() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("[URL=client://");
```
Comment (Firedroide) [inline src/main/java/com/github/theholywaffle/teamspeak3/api/wrapper/Client.java]: `append` can be chained, e.g. `sb.append("name=").append(clientName).append(";");` It might be possible to group a few of these statements together so this method doesn't get so bulky.

File: src/main/java/com/github/theholywaffle/teamspeak3/api/wrapper/Client.java [hunk 6]
```diff
@@ -28,174 +28,201 @@
...
+    public int getChannelId() {
+        return getInt(ClientProperty.CID);
+    }
+
+    /**
+     * Utility method. This method will return a client uri.
+     * A client uri can be used to reference a client.
+     * Example: [URL=client://5/thisisanuid=~Jesus]Jesus[/URL]
+     *
+     * @return clientUri
```
Comment (Firedroide) [inline src/main/java/com/github/theholywaffle/teamspeak3/api/wrapper/Client.java]: `@return an URI that links to this client`

URI should be capitalized in the JavaDoc above, as well :stuck_out_tongue:

**Comment by Scrumplex** [comment]
Tabs vs. Spaces

**Comment by Firedroide** [comment]
Kinda, yeah ^^
We decided to use tabs in this project and because changing all source code over just whitespace is a dumb idea (and tabs > spaces anyways), we'll stick with that decision :smile:

By the way, I've checked out these URLs, and I've noticed there are some problems when it comes to client nicknames.
First of all, it doesn't seem like encoding the nicknames with `URLEncoder` is the right thing to do. These URLs only work inside the teamspeak client anyways and special characters seem to work just fine in ts3 clients. The few exceptions being spaces, angle brackets and slashes, of course, which break the BB-code / URL.
What is more, the client doesn't decode URLs encoded with `URLEncoder` correctly way when displaying the link.

Luckily, the client URLs still work without a nickname or with a completely wrong one.
So it might be a better idea to just leave out the nickname for good if the only thing it does is cause problems. But I'd love to hear your input on this.

**Comment by Scrumplex** [comment]
I'll leave it because I tried one of those "Cool name generators" where text looks like a mess and TS3 encoded and decoded these characters.

**Comment by Firedroide** [comment]
Thanks for the PR, merged! :smile:

By the way, seems like GitHub has added a new feature called "Squash and merge". It does just that, squashes all commits of this PR into a single one and pseudo-merges the branches. I've wanted something like that for years!

image

It's so fancy! :smile:

**Comment by Scrumplex** [comment]
`Scrumplex committed with Firedroide` does sound weird.

**Comment by Firedroide** [comment]
True, but it gives attribution to the person who created the PR and not just the one who "created the commit" from the git point-of-view - I don't really care how they phrase it :stuck_out_tongue: 

And the feature it saves me from having to ask everyone "oh, and by the way, don't forget to squash your commits! Here's how to do that..." ^^
